src/main/java/org/wildfly/naming/client/NamingProviderFactory.java

```diff
@@ -47,10 +48,10 @@
     /**
      * Create the naming provider instance for a provider URI.
      *
-     * @param providerUri the URI from the {@link Context#PROVIDER_URL} environment property, or {@code null} if none was given
      * @param env a copy of the environment which may be consumed directly by the provider (not {@code null})
+     * @param providerUris the URIs from the {@link Context#PROVIDER_URL} environment property, or {@code null} if none were given
      * @return the root context (must not be {@code null})
      * @throws NamingException if the root context creation failed for some reason
      */
-    NamingProvider createProvider(URI providerUri, FastHashtable<String, Object> env) throws NamingException;
+    T createProvider(FastHashtable<String, Object> env, URI... providerUris) throws NamingException;
```

Comment:
I think making this generic is not necessary, if I understand what you are trying to do.  You can instead have the interface return NamingProvider while still having RemoteNamingProviderFactory return a RemoteNamingProvider, using covariant return type overriding.